ODIN_II/SRC/netlist_create_from_ast.cpp

```diff
@@ -3225,106 +3225,127 @@ int alias_output_assign_pins_to_inputs(char_list_t* output_list, signal_list_t*
-    signal_list_t* out_1 = NULL;
-
-    for (long j = 0; j < gate->children[0]->num_children; j++) {
-        /* free the previous list since it was not the last itteration */
-        if (out_1 != NULL) {
-            free_signal_list(out_1);
-            out_1 = NULL;
+void connect_gate_instantiation(ast_node_t* gate, char* instance_name_prefix, sc_hierarchy* local_ref) {
+    // must at least have a name
+    oassert(gate->num_children >= 1);
+
+    signal_list_t** input_list = NULL;
+    size_t input_size = 0;
+
+    signal_list_t** output_list = NULL;
+    size_t output_size = 0;
+
+    switch (gate->types.operation.op) {
+        case NOTIF0: //fallthrough
+        case NOTIF1: //fallthrough
+        case BUFIF0: //fallthrough
+        case BUFIF1: {
+            if (gate->num_children != 4) {
+                //error expected 1 output and 2 input
+                oassert(false);
+            }
+            input_size = 2;
+            output_size = 1;
+
+            break;
         }
+        case BITWISE_NOT: //fallthrough
+        case BUF: {
+            if (gate->num_children < 3) {
+                //error expected at least n+ output and 1 input
+                oassert(false);
+            }
+            input_size = 1;
+            output_size = gate->num_children - 2;
 
-        ast_node_t* gate_instance = gate->children[0]->children[j];
+            break;
+        }
+        case BITWISE_AND:  //fallthrough
+        case BITWISE_NAND: //fallthrough
+        case BITWISE_NOR:  //fallthrough
+        case BITWISE_OR:   //fallthrough
+        case BITWISE_XNOR: //fallthrough
+        case BITWISE_XOR: {
+            if (gate->num_children < 3) {
+                //error expected at least 1 output and n+ input
+                oassert(false);
+            }
+            input_size = gate->num_children - 2;
+            output_size = 1;
 
-        if (gate_instance->children[3] == NULL) {
-            /* IF one input gate */
+            break;
+        }
+        default: {
+            oassert(false
+                    && "Unrecognized type of gate passed");
+            break;
+        }
+    }
 
-            /* process the signal for the input gate */
-            signal_list_t* in_1 = netlist_expand_ast_of_module(&(gate_instance->children[2]), instance_name_prefix, local_ref);
-            /* process the signal for the input ga$te */
-            out_1 = create_output_pin(gate_instance->children[1], instance_name_prefix, local_ref);
+    output_list = (signal_list_t**)vtr::calloc(output_size, sizeof(signal_list_t*));
+    input_list = (signal_list_t**)vtr::calloc(input_size, sizeof(signal_list_t*));
 
-            oassert(in_1 != NULL);
-            oassert(out_1 != NULL);
+    // skip symbol
+    size_t child_id = 1;
+    for (size_t i = 0; i < output_size; i += 1) {
+        output_list[i] = create_output_pin(gate->children[child_id], instance_name_prefix, local_ref);
```

Comment:
this is the only reason the children cannot be done in the netlist_create_from_ast, It we can flag outputs from instances as such in the types, they could be recognized and handled there rather, reducing a lot of clutter.